src/core/bpf-util.c

```diff
@@ -0,0 +1,37 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#if BPF_FRAMEWORK
+
+#include "bpf-dlopen.h"
+#include "bpf-util.h"
+#include "cgroup-util.h"
+#include "log.h"
+
+bool bpf_supported(void) {
+        static int supported = -1;
+        int r;
+
+        if (supported >= 0)
+                return supported;
+
+        r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
+        if (r < 0) {
+                log_warning_errno(r, "Can't determine whether the unified hierarchy is used: %m");
+                return (supported = false);
+        }
+
+        if (r == 0) {
+                log_info_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                               "Not running with unified cgroup hierarchy, BPF cgroup filters are not supported");
```

Comment:
Copy paste error with the log messages?

Reply:
Hmm, can you be more specific?

Reply:
> BPF cgroup filters are not supported

This looked specific for one use case of BPF but I might understand BPF wrong

Reply:
"BPF cgroup filters" was supposed to be a generic term that covers the types of filters we attach… I think it's OK,  but suggestions are welcome.

Reply:
I got tripped up by the function being named bpf_supported(). I figure we might use bpf for something different than cgroup filters in the future and then the message would be wrong if we also use bpf_supported() there to check for bpf support. 

So I'd say either the function should be renamed to something more specific to indicate it's checking more than just whether bpf is supported or the error message should be updated to say we don't support any BPF features unless we're using the unified hierarchy, so something like "Not running with unified cgroup hierarchy, disabling BPF features"

Reply:
I renamed the function to `cgroup_bpf_supported()` and adjusted the texts a bit.